flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroFileFormatFactory.java

```diff
@@ -90,6 +135,112 @@ public String factoryIdentifier() {
         return options;
     }
 
+    private static class AvroGenericRecordBulkFormat
+            extends AbstractAvroBulkFormat<GenericRecord, RowData, FileSourceSplit> {
+
+        private static final long serialVersionUID = 1L;
+
+        // all physical fields in the source schema
+        private final DataType physicalDataType;
+        // projected physical fields
+        private final DataType producedDataType;
+        TypeInformation<RowData> producedTypeInfo;
+        private final List<String> partitionKeys;
+        private final String defaultPartitionValue;
+
+        private transient AvroToRowDataConverters.AvroToRowDataConverter converter;
+        private transient GenericRecord reusedAvroRecord;
+        private transient GenericRowData reusedRowData;
+        // we should fill i-th field of reusedRowData with
+        // readerRowTypeIndex[i]-th field of reusedAvroRecord
+        private transient int[] readerRowTypeIndex;
+
+        public AvroGenericRecordBulkFormat(
+                DynamicTableSource.Context context,
+                DataType physicalDataType,
+                DataType producedDataType,
+                List<String> partitionKeys,
+                String defaultPartitionValue) {
+            // partition keys are stored in file paths, not in avro file contents
+            super(
+                    AvroSchemaConverter.convertToSchema(
+                            getNotNullRowTypeWithExclusion(producedDataType, partitionKeys)));
+            this.physicalDataType = physicalDataType;
+            this.producedDataType = producedDataType;
+            this.producedTypeInfo = context.createTypeInformation(producedDataType);
+            this.partitionKeys = partitionKeys;
+            this.defaultPartitionValue = defaultPartitionValue;
+        }
+
+        @Override
+        protected void open(FileSourceSplit split) {
+            RowType readerRowType = getNotNullRowTypeWithExclusion(producedDataType, partitionKeys);
+
+            converter = AvroToRowDataConverters.createRowConverter(readerRowType);
+            reusedAvroRecord = new GenericData.Record(readerSchema);
+
+            List<String> physicalFieldNames = DataType.getFieldNames(physicalDataType);
+            int[] selectFieldIndices =
+                    DataType.getFieldNames(producedDataType).stream()
+                            .mapToInt(physicalFieldNames::indexOf)
+                            .toArray();
```

Comment:
Same as above, no need to project fields etc. Just use the `producedDataType` provided by you in `createRuntimeDecoder`